cdap-master/src/main/java/co/cask/cdap/data/tools/RemoteDatasetServiceManager.java

```
@@ -0,0 +1,131 @@
+ */
+public class RemoteDatasetServiceManager extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(UpgradeTool.class);
+
+  private final DatasetService datasetService;
+  private final ZKClientService zkClientService;
+  private final DatasetFramework datasetFramework;
+  private final DatasetOpExecutorService datasetOpExecutorService;
+
+  @Inject
+  RemoteDatasetServiceManager(CConfiguration cConf, Configuration hConf) {
+    Injector injector = initInjector(cConf, hConf);
+    this.datasetService = injector.getInstance(DatasetService.class);
+    this.zkClientService = injector.getInstance(ZKClientService.class);
+    this.datasetFramework = injector.getInstance(DatasetFramework.class);
+    this.datasetOpExecutorService = injector.getInstance(DatasetOpExecutorService.class);
+
+  }
+
+  public DatasetFramework getDSFramework() {
+    return datasetFramework;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    if (!zkClientService.isRunning()) {
+      zkClientService.startAndWait();
+    }
+    datasetOpExecutorService.startAndWait();
+    datasetService.startAndWait();
+
+    // wait 5 minutes for DatasetService to start up
+    Tasks.waitFor(true, new Callable<Boolean>() {
+      @Override
+      public Boolean call() throws Exception {
+        try {
+          getDSFramework().getInstances(Id.Namespace.DEFAULT);
+          return true;
+        } catch (ServiceUnavailableException sue) {
+          return false;
+        }
+      }
+    }, 5, TimeUnit.MINUTES, 10, TimeUnit.SECONDS);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    try {
+      datasetService.stopAndWait();
+      datasetOpExecutorService.startAndWait();
+      zkClientService.stopAndWait();
+    } catch (Throwable e) {
+      LOG.error("Exception while trying to stop services running for {}",
+                RemoteDatasetServiceManager.class.getSimpleName(), e);
+      throw Throwables.propagate(e);
```

hmmm, should we propagate a runtime error here? you're just trying to stop services.